src/core/sadder.rs

```diff
@@ -0,0 +1,413 @@
+use crate::core::{
+    common::{loads, sizeify, sniff, Ids, Version, CURRENT_VERSION},
+    matter::{tables as matter, Matter},
+    saider::Saider,
+};
+use crate::data::{Data, Value};
+use crate::error::{err, Error, Result};
+
+#[derive(Debug, Clone, PartialEq)]
+struct ExhaleResult {
+    raw: Vec<u8>,
+    ident: String,
+    kind: String,
+    ked: Value,
+    version: Version,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+struct InhaleResult {
+    ked: Value,
+    ident: String,
+    kind: String,
+    version: Version,
+    size: u32,
+}
+
+fn inhale(raw: &[u8]) -> Result<InhaleResult> {
+    let result = sniff(raw)?;
+
+    if result.version != *CURRENT_VERSION {
+        return err!(Error::Value(format!(
+            "unsupported version = {:?}, current version = {:?}",
+            result.version, CURRENT_VERSION
+        )));
+    }
+
+    let ked = loads(raw, Some(result.size), Some(&result.kind))?;
+
+    Ok(InhaleResult {
+        ked,
+        ident: result.ident,
+        kind: result.kind,
+        version: result.version,
+        size: result.size,
+    })
+}
+
+fn exhale(ked: &Value, kind: Option<&str>) -> Result<ExhaleResult> {
+    let result = sizeify(ked, kind)?;
+    Ok(ExhaleResult {
+        raw: result.raw,
+        ident: result.ident,
+        ked: result.ked,
+        kind: result.kind,
+        version: result.version,
+    })
+}
+
+pub(crate) trait Sadder: Default + Clone {
```

Comment:
I would love to see a few header comments for all "trait" definitions. 

What is a Sadder again? What are the etymological roots?

Reply:
SAD = Sad Addressing Data I believe but that is from memory